### PR TITLE
CUDA softmax kernel and broadcast kernel support for multiple non-consecutive axes

### PR DESCRIPTION
**Pull request by csullivan**
This PR introduces fused elementwise+collective kernels. op::Softmax is implemented in the GPU via two of these kernels: 
```
1. elementwise exponentiation fused with sum reduce
2. elementwise division fused with broadcast
```

Features:
* Adds softmax cuda kernel to handle activation along arbitrary axes.
* Adds a few new softmax tests
* Adds broadcast to the CUDAEmitter with support for all axis broadcasting (coordinate transformation function is shared in softmax and broadcast)

Limitations:
* The softmax kernel relies on atomic updates to global memory which can create contention. I began work on a hierarchical reduction in cs/gpu_softmax_cuda_shfl, but decided to postpone further efforts until later.

### CODE REVIEW

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 1]
```diff
@@ -977,12 +1174,24 @@ std::string runtime::gpu::CUDAEmitter::include_helpers()
...
     ss << R"(
+#define WARP_SIZE 32
 #define __ballot_sync(mask, predicate) __ballot(predicate)
 #define __shfl_down_sync(mask, val, delta, width) __shfl_down(val, delta, width)
 #define __shfl_xor_sync(mask, val, laneMask, width) __shfl_xor(val, laneMask, width)
 )";
 #endif
 
+    // add modern type definitions
+    ss << "typedef signed char int8_t;\n";
```
Comment (rkimballn1) [inline src/ngraph/runtime/gpu/cuda_emitter.cpp]: These are defined in header files. #include <inttypes> or whatever

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 2]
```diff
@@ -23,15 +23,15 @@ using namespace ngraph;
 void runtime::gpu::CudaKernelBuilder::get_elementwise_op(codegen::CodeWriter& writer,
                                                          const std::string& name,
                                                          const std::string& op,
-                                                         const std::vector<std::string>& data_types)
+                                                         const std::vector<std::string>& dtypes)
```
Comment (rkimballn1) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: It would have been better to do the rename in a separate PR. This just adds noise.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 3]
```diff
@@ -372,20 +383,104 @@ void runtime::gpu::CudaKernelBuilder::get_replace_slice_op(
...
+    codegen::CodeWriter& writer,
+    std::string i_thread_index,
+    std::string i_strides,
+    std::string i_stride_magic,
+    std::string i_stride_shift,
+    std::string i_reduced_strides,
+    std::string o_coordinates,
+    int rank)
+{
+    /***
```
Comment (rkimballn1) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Single line comments please

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 4]
```diff
@@ -372,20 +383,104 @@ void runtime::gpu::CudaKernelBuilder::get_replace_slice_op(
...
+            auto reduced_idx = reduce_coordinate_transform_helper(writer,
+                                                                  "tid",
+                                                                  "strides",
+                                                                  "stride_magic",
+                                                                  "stride_shift",
+                                                                  "reduced_strides",
+                                                                  "coordinate",
+                                                                  rank);
+
+            // TODO: mediate atomic memory access contention
```
Comment (rkimballn1) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Please file a ticket to track this todo

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp (hunk 2 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: `daya_types` should be good?

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 5]
```diff
@@ -937,6 +937,203 @@ size_t runtime::gpu::CUDAEmitter::build_replace_slice(const GPURuntimeContext* c
...
+size_t runtime::gpu::CUDAEmitter::build_softmax(const GPURuntimeContext* ctx,
+                                                const std::array<std::string, 2>& dtypes,
+                                                GPUShape tensor_shape,
+                                                const AxisSet& reduce_axes)
+{
+    // assumes NC{d1,...,dn} format
+    std::string kernel_name = "softmax_" + join(dtypes, "_");
+    std::replace(kernel_name.begin(), kernel_name.end(), ' ', '_');
+
+    // TODO before PR - come back and check if shape is necessary, could be just rank is needed
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/cuda_emitter.cpp]: checked? remove?

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp (hunk 2 above)
Comment (csullivan) [reply src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: It just needs to be consistent, `dtypes` were used in some places while `data_types` were used elsewhere. I chose dtypes because it is shorter and more commonly seen in other frameworks (mxnet, neon, tf, etc). I can remove it from this PR if desired.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp (hunk 1 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: we can use a .h file for cuda header, than pass to nvrtc.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 6]
```diff
@@ -1031,6 +1239,15 @@ __device__ __forceinline__ float  load(const float*  __restrict__ in, int i=0, b
     }
     return v;
 }
+__device__ __forceinline__ int64_t  load(const int64_t*  __restrict__ in, int i=0, bool b=true)
+{
+    int64_t v = 0.0f;
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/cuda_emitter.cpp]: 0 is OK.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 7]
```diff
@@ -372,6 +384,89 @@ void runtime::gpu::CudaKernelBuilder::get_replace_slice_op(
...
+                                                                         "stride_shift",
+                                                                         "reduced_strides",
+                                                                         "coordinate",
+                                                                         rank);
+
+            // TODO: mediate atomic memory access contention
+            writer << data_types[0] << " val = expf(load(in, tid));\n";
+            writer << "atomicAdd(&out[" << reduced_idx << "], val);\n";
+            writer << "__threadfence();\n";
+            writer << data_types[1] << " sum = out[" << reduced_idx << "];\n";
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: This will only work when all data of reduced axis fall in the same block.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.hpp [hunk 8]
```diff
@@ -63,23 +64,39 @@ namespace ngraph
...
                 static void get_device_helper(codegen::CodeWriter& writer,
                                               const std::string& name,
                                               const std::string& math_kernel,
                                               const std::vector<std::string>& data_types);
 
                 static void add_pod_typedefs(codegen::CodeWriter& writer);
+
+                /// \brief Given kernel input variables i_* produce register variables o_coordinates{i}
+                ///        and return string name of integer index into reduced array
+                static std::string reduce_coordinate_transform_helper(codegen::CodeWriter& writer,
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.hpp]: nice.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp (hunk 7 above)
Comment (csullivan) [reply src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Softmax is now implemented via fused elementwise collective operations. See gpu_emitter.cpp.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 9]
```diff
@@ -57,28 +57,131 @@ void runtime::gpu::CudaKernelBuilder::get_elementwise_op(codegen::CodeWriter& wr
...
+                }
+                writer << "]";
+            }
+            writer << ");\n";
+
+            // global collective reduce or broadcast
+            if (reduce_op != "")
+            {
+                // TODO: mediate atomic memory access contention
+                writer << reduce_op << "(&out0[" << reduced_idx << "], output);\n";
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: This is not the best way to do reduce. we can visit it later.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 10]
```diff
@@ -937,6 +1073,104 @@ size_t runtime::gpu::CUDAEmitter::build_replace_slice(const GPURuntimeContext* c
...
+    size_t idx_reduced_strides =
+        allocator.reserve_argspace(reduced_strides.data(), reduced_strides.size() * sizeof(int));
+
+    // TODO: blending factors are not currently implemented
+    float alpha = 1.0f;
+    float beta = 0.0f;
+
+    int nthreads = static_cast<int>(shape_size(result_shape));
+
+    std::unique_ptr<gpu::primitive> softmax(new gpu::primitive{[=](void** inputs,
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/cuda_emitter.cpp]: It's a little bit confuse when build broadcast with softmax.

Comment (csullivan) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: Fixed.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp (hunk 9 above)
Comment (csullivan) [reply src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Correct, I added a ticket related to this last week.

File: src/ngraph/runtime/gpu/cuda_emitter.hpp [hunk 11]
```diff
@@ -71,13 +71,36 @@ namespace ngraph
                                            GPUShape reduce_window_shape,
                                            GPUShape reduce_window_strides);
 
-                template <typename T>
+                template <typename ELEMENTWISE_OP_TYPE, typename REDUCE_OP_TYPE = ngraph::op::Nop>
```
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cuda_emitter.hpp]: Refactored as suggested.

**Comment by csullivan** [comment]
TESTNOW
